File: JavaScript/19-Remove Nth Node From End of List.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
var removeNthFromEnd = function(head, n) {
    let first = head, second = head;
    for (let i=0; i<n; i++){
        first = first.next;
    }
    
    if (!first) return head.next;
    
    while (first.next){
        first = first.next;
        second = second.next;
    }
    
    second.next = second.next.next;
    
    return head;
    
};

File: C++/19-Remove Nth Node From End of List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        int cnt = 0;
        ListNode* tmp = head;
        while (tmp) {
            cnt++;
            tmp = tmp->next;
        }
        
        int ind = cnt-n;
        ListNode* prev = new ListNode(-1);
        prev->next = head;
        
        int nextIdx = 0;
        ListNode* running = prev;
        while (running) {
            if (nextIdx == ind) {
                running->next = running->next->next;
            }
            running = running->next;
            nextIdx++;
        }
        
        return prev->next;
    }
};